### PR TITLE
KY 93646 - Added terraform support for azure connection

### PR DESCRIPTION
**Pull request by udita02**

### CODE REVIEW

File: internal/provider/connections/resource_azure_connection.go [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+	Options:
+	
+		AzureCloud
+		AzureChinaCloud
+		AzureUSGovernment
+		AzureStack
+`
+)
+
+func NewResourceCMAzureConnection() resource.Resource {
```
Comment (om-shrirao) [inline internal/provider/connections/resource_azure_connection.go:37]: dont add CMAzureConnection, don't add prefix cm, remov it from everywhere in this file

File: internal/provider/connections/resource_azure_connection.go [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+	return &resourceCMAzureConnection{}
+}
+
+type resourceCMAzureConnection struct {
```
Comment (om-shrirao) [inline internal/provider/connections/resource_azure_connection.go:41]: same here cm prefix

File: internal/provider/connections/resource_azure_connection.go [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+				Optional:    true,
+				Description: "Description about the connection.",
+			},
+			"external_certificate_used": schema.BoolAttribute{
+				Optional:    true,
+				Description: "true if the certificate associated with the connection is generated externally, false otherwise.",
+			},
+			"is_certificate_used": schema.BoolAttribute{
+				Optional:    true,
+				Description: "User has the option to choose the Certificate Authentication method instead of Client Secret for Azure Cloud connection. In order to use the Certificate, set it to true. Once the connection is created, in the response user will get a certificate. By default, the certificate is valid for 2 Years. User can update the certificate in the existing connection by setting it to true in Update (PATCH) API call.",
```
Comment (om-shrirao) [inline internal/provider/connections/resource_azure_connection.go:109]: no need of "setting it to true in Update (PATCH) API call"

Comment (om-shrirao) [reply internal/provider/connections/resource_azure_connection.go:109]: change it to something else

File: internal/provider/connections/resource_azure_connection.go [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+				Optional:    true,
+				Description: "Azure stack server certificate",
+			},
+			"cert_duration": schema.Int64Attribute{
+				Optional:    true,
+				Description: "Duration in days for which the azure certificate is valid, default (730 i.e. 2 Years).",
+			},
+			"certificate": schema.StringAttribute{
+				Optional:    true,
+				Description: "User has the option to upload external certificate for Azure Cloud connection. This option cannot be used with option is_certificate_used and client_secret.User first has to generate a new Certificate Signing Request (CSR) in POST /v1/connectionmgmt/connections/csr. The generated CSR can be signed with any internal or external CA. The Certificate must have an RSA key strength of 2048 or 4096. User can also update the new external certificate in the existing connection in Update (PATCH) API call. Any unused certificate will automatically deleted in 24 hours.",
```
Comment (om-shrirao) [inline internal/provider/connections/resource_azure_connection.go:89]: no need of: "Update (PATCH) API call", change it to something else

File: internal/provider/connections/resource_azure_connection.go [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+	}
+}
+
+// Create creates the resource and sets the initial Terraform state.
+func (r *resourceCMAzureConnection) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	id := uuid.New().String()
+	tflog.Trace(ctx, common.MSG_METHOD_START+"[resource_azure_connection.go -> Create]["+id+"]")
+
+	// Retrieve values from plan
+	var plan CMAzureConnectionTFSDK
```
Comment (om-shrirao) [inline internal/provider/connections/resource_azure_connection.go:163]: no cm prefix in both struct too

File: internal/provider/connections/resource_azure_connection.go [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+	plan.Account = types.StringValue(gjson.Get(response, "account").String())
+	plan.UpdatedAt = types.StringValue(gjson.Get(response, "updatedAt").String())
+	plan.CreatedAt = types.StringValue(gjson.Get(response, "createdAt").String())
+	plan.Category = types.StringValue(gjson.Get(response, "category").String())
+	plan.Service = types.StringValue(gjson.Get(response, "service").String())
+	plan.ResourceURL = types.StringValue(gjson.Get(response, "resource_url").String())
+	plan.LastConnectionOK = types.BoolValue(gjson.Get(response, "last_connection_ok").Bool())
+	plan.LastConnectionError = types.StringValue(gjson.Get(response, "last_connection_error").String())
+	plan.LastConnectionAt = types.StringValue(gjson.Get(response, "last_connection_at").String())
+
```
Comment (om-shrirao) [inline internal/provider/connections/resource_azure_connection.go:289]: Any other computed values which needs to be set, which is given in response?

File: internal/provider/connections/resource_azure_connection.go [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+	tflog.Debug(ctx, fmt.Sprintf("Response: %s", response))
```
Comment (om-shrirao) [inline internal/provider/connections/resource_azure_connection.go:290]: redundant log for response

File: internal/provider/connections/schema_connections.go [hunk 8]
```diff
@@ -122,6 +122,56 @@ type CMCreateConnectionResponseCommon struct {
 	LastConnectionAt    string `json:"last_connection_at"`
 }
 
+type CMAzureConnectionTFSDK struct {
```
Comment (om-shrirao) [inline internal/provider/connections/schema_connections.go:125]: remove prefix CM from all struct

File: internal/provider/provider.go [hunk 9]
```diff
@@ -334,5 +335,6 @@ func (p *ciphertrustProvider) Resources(_ context.Context) []func() resource.Res
 		cm.NewResourceCMTrialLicense,
 		cm.NewResourceCMPrometheus,
 		connections.NewResourceGCPConnection,
+		connections.NewResourceCMAzureConnection,
```
Comment (om-shrirao) [inline internal/provider/provider.go:338]: remove CM prefix

File: sample-scripts/connections/azure/main.tf [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  password = "Zxcv1234."
+  bootstrap = "no"
+}
+
+resource "ciphertrust_azure_connection" "azure_connection" {
+  name        = "azure-connection"
+  products = [
+    "cckm"
+  ]
+   client_secret="3bf0dbe6-a2c7-431d-9a6f-4843b74c71285nfjdu2"
```
Comment (om-shrirao) [inline sample-scripts/connections/azure/main.tf:22]: format it properly

File: internal/provider/connections/resource_azure_connection.go [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+			"products": schema.ListAttribute{
+				ElementType: types.StringType,
+				Optional:    true,
+				Description: productsDescription,
+			},
+			"resource_manager_url": schema.StringAttribute{
+				Optional:    true,
+				Description: "Azure stack resource manager URL.",
+			},
+			"vault_resource_url": schema.StringAttribute{
```
Comment (VJ-VijayThales) [inline internal/provider/connections/resource_azure_connection.go:138]: isnt these parameters computed ??

**Comment by VJ-VijayThales** [comment]
branch has merge conflicts!!

**Review by VJ-VijayThales — APPROVED** [review]

File: internal/provider/connections/resource_azure_connection.go (hunk 11 above)
Comment (om-shrirao) [reply internal/provider/connections/resource_azure_connection.go:138]: +1

Comment (udita02) [reply internal/provider/connections/resource_azure_connection.go:138]: Nope, This parameter is provided by user while creating the resource

**Review by om-shrirao — APPROVED** [review]

**Pull request merged**